### PR TITLE
Centralize informer creation in agent and server

### PR DESCRIPTION
**Pull request by mcbenjemaa**
## Description

as requested https://github.com/prometheus-operator/prometheus-operator/issues/5454
This PR created the main common informers and passed them to the server and agent.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [X] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

Fixes https://github.com/prometheus-operator/prometheus-operator/issues/5454

### CODE REVIEW

**Comment by mcbenjemaa** [comment]
Do you agree with what I implemented, or do you have any better options?

**Review by JoaoBraveCoding — COMMENTED** [review]
Some minor comments but definitely in a good direction! Good work!

File: cmd/operator/main.go [hunk 1]
```diff
@@ -429,6 +446,143 @@ func Main() int {
 	return 0
 }
 
+func initCommonConfig(c operator.Config) (*prompkg.CommonConfig, error) {
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply cmd/operator/main.go:449]: make sense

File: cmd/operator/main.go [hunk 2]
```diff
@@ -429,6 +446,143 @@ func Main() int {
...
+		informers.NewMonitoringInformerFactories(
+			c.Namespaces.PrometheusAllowList,
+			c.Namespaces.DenyList,
+			cm.MClient,
+			prompkg.ResyncPeriod,
+			func(options *metav1.ListOptions) {
+				options.LabelSelector = c.PromSelector
+			},
+		),
+		monitoringv1alpha1.SchemeGroupVersion.WithResource(monitoringv1alpha1.PrometheusAgentName),
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply cmd/operator/main.go:483]: Or having both 
`promInfs` and `promAgInfs`

File: cmd/operator/main.go [hunk 3]
```diff
@@ -429,6 +446,143 @@ func Main() int {
...
+	cm.MClient, err = monitoringclient.NewForConfig(cm.ClusterConfig)
+	if err != nil {
+		return nil, pkgerrors.Wrap(err, "instantiating monitoring client failed")
+	}
+
+	secretListWatchSelector, err := fields.ParseSelector(c.SecretListWatchSelector)
+	if err != nil {
+		return nil, pkgerrors.Wrap(err, "can not parse secrets selector value")
+	}
+
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply cmd/operator/main.go:471]: What do you mean?

**Comment by mcbenjemaa** [comment]
@JoaoBraveCoding can you check

File: pkg/prometheus/agent/operator.go [hunk 4]
```diff
@@ -83,39 +79,27 @@ type Operator struct {
...
-
-	mclient, err := monitoringclient.NewForConfig(cfg)
-	if err != nil {
-		return nil, errors.Wrap(err, "instantiating monitoring client failed")
-	}
+func New(ctx context.Context, conf operator.Config, logger log.Logger, r prometheus.Registerer, cm *prompkg.CommonConfig) (*Operator, error) {
 
 	if _, err := labels.Parse(conf.PromSelector); err != nil {
 		return nil, errors.Wrap(err, "can not parse prometheus-agent selector value")
 	}
```
Comment (JoaoBraveCoding) [inline pkg/prometheus/agent/operator.go:86]:
```suggestion
```

File: pkg/prometheus/server/operator.go [hunk 5]
```diff
@@ -91,36 +87,17 @@ type Operator struct {
...
 	}
 
-	mclient, err := monitoringclient.NewForConfig(cfg)
-	if err != nil {
-		return nil, errors.Wrap(err, "instantiating monitoring client failed")
-	}
-
 	if _, err := labels.Parse(conf.PromSelector); err != nil {
 		return nil, errors.Wrap(err, "can not parse prometheus selector value")
 	}
```
Comment (JoaoBraveCoding) [inline pkg/prometheus/server/operator.go:100]:
```suggestion
```

File: pkg/prometheus/operator.go [hunk 6]
```diff
@@ -106,3 +135,158 @@ func ValidateRemoteWriteSpec(spec monitoringv1.RemoteWriteSpec) error {
...
+	cm.MClient, err = monitoringclient.NewForConfig(cm.ClusterConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "instantiating monitoring client failed")
+	}
+
+	secretListWatchSelector, err := fields.ParseSelector(c.SecretListWatchSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "can not parse secrets selector value")
+	}
+
```
Comment (JoaoBraveCoding) [inline pkg/prometheus/operator.go:162]:
```suggestion
	if _, err := labels.Parse(conf.PromSelector); err != nil {
		return nil, errors.Wrap(err, "can not parse prometheus selector value")
	}
```

**Review by JoaoBraveCoding — COMMENTED** [review]
Clarifying https://github.com/prometheus-operator/prometheus-operator/pull/5499#discussion_r1171108419 but overall lgtm, lovely work!

File: pkg/prometheus/agent/operator.go (hunk 4 above)
Comment (mcbenjemaa) [reply pkg/prometheus/agent/operator.go:86]: I don't understand?

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm

File: pkg/prometheus/operator.go [hunk 7]
```diff
@@ -106,3 +136,160 @@ func ValidateRemoteWriteSpec(spec monitoringv1.RemoteWriteSpec) error {
...
+	secretListWatchSelector, err := fields.ParseSelector(c.SecretListWatchSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "can not parse secrets selector value")
+	}
+	if _, err := labels.Parse(c.PromSelector); err != nil {
+		return nil, errors.Wrap(err, "can not parse prometheus selector value")
+	}
+	// init promInfs
+	cm.PromInfs, err = informers.NewInformersForResource(
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:167]: this isn't needed by the PrometheusAgent controller.

File: pkg/prometheus/operator.go [hunk 8]
```diff
@@ -106,3 +136,160 @@ func ValidateRemoteWriteSpec(spec monitoringv1.RemoteWriteSpec) error {
...
+			},
+		),
+		monitoringv1.SchemeGroupVersion.WithResource(monitoringv1.PrometheusName),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating prometheus informers")
+	}
+
+	// init promAgInfs
+	cm.PromAgInfs, err = informers.NewInformersForResource(
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:184]: same here, the Prometheus controller doesn't need it.

File: pkg/prometheus/operator.go [hunk 9]
```diff
@@ -19,17 +19,47 @@ import (
...
+	MClient       *monitoringclient.Clientset
+
+	PromInfs   *informers.ForResource
+	PromAgInfs *informers.ForResource
+	SmonInfs   *informers.ForResource
+	PmonInfs   *informers.ForResource
+	ProbeInfs  *informers.ForResource
+	CmapInfs   *informers.ForResource
+	SecrInfs   *informers.ForResource
+	SsetInfs   *informers.ForResource
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:60]: I think that it makes sense to have a struct regrouping informers for servicemonitors, podmonitors, probes, confimaps, statefulsets and secrets .
On the other side, prometheus and prometheusagents informers need to live in their respective controllers.

File: pkg/prometheus/operator.go [hunk 10]
```diff
@@ -19,17 +19,47 @@ import (
...
+	ResyncPeriod = 5 * time.Minute
 )
 
 var prometheusKeyInShardStatefulSet = regexp.MustCompile("^(.+)/prometheus-(.+)-shard-[1-9][0-9]*$")
 var prometheusKeyInStatefulSet = regexp.MustCompile("^(.+)/prometheus-(.+)$")
 
+type CommonConfig struct {
+	ClusterConfig *rest.Config
+	KClient       *kubernetes.Clientset
+	MClient       *monitoringclient.Clientset
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:51]: I'm not sure about the benefit of having these fields. The struct feels a bit like a mixed bag of unrelated things and I'd focus on informers first.

File: pkg/prometheus/operator.go [hunk 11]
```diff
@@ -106,3 +136,160 @@ func ValidateRemoteWriteSpec(spec monitoringv1.RemoteWriteSpec) error {
...
+		return nil, errors.Wrap(err, "instantiating monitoring client failed")
+	}
+
+	secretListWatchSelector, err := fields.ParseSelector(c.SecretListWatchSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "can not parse secrets selector value")
+	}
+	if _, err := labels.Parse(c.PromSelector); err != nil {
+		return nil, errors.Wrap(err, "can not parse prometheus selector value")
+	}
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply pkg/prometheus/operator.go:165]: `labels.Selector` is an interface so i will create a new type PromSelector of type string.

File: pkg/prometheus/operator.go (hunk 10 above)
Comment (mcbenjemaa) [reply pkg/prometheus/operator.go:51]: But those clients needed to initialize the informers, if not moving them here..
We will need to duplicate them (init them twice).
I don't think this is a good idea.

File: pkg/prometheus/operator.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply pkg/prometheus/operator.go]: this is done, can you check what I changed?

**Comment by mcbenjemaa** [comment]
I have also move scrapeconfig informer to commonConfig

File: pkg/prometheus/operator.go (hunk 10 above)
Comment (JoaoBraveCoding) [reply pkg/prometheus/operator.go:53]: I agree with @simonpasquier that I'm feeling a trend with this and this [1], that we will end up with multiple structs that will have the clients duplicated. Now I'm not sure what @simonpasquier has in mind, but maybe we could have a struct called `Clients` that would initialise and pass to the Operators which would then, in turn, pass them to functions like `Process` or `InitCommonConfig`. But let's see what @simonpasquier suggests

[1] https://github.com/prometheus-operator/prometheus-operator/pull/5493

**Review by JoaoBraveCoding — COMMENTED** [review]
Just a comment on the struct but apart from that I don't have anything else

File: pkg/prometheus/operator.go (hunk 10 above)
Comment (mcbenjemaa) [reply pkg/prometheus/operator.go:53]: I'm not sure, if I understand this well.

File: pkg/prometheus/agent/operator.go [hunk 12]
```diff
@@ -177,62 +119,32 @@ func New(ctx context.Context, conf operator.Config, logger log.Logger, r prometh
...
-			nil,
-		),
-		monitoringv1.SchemeGroupVersion.WithResource(monitoringv1.ProbeName),
-	)
-	if err != nil {
-		return nil, errors.Wrap(err, "error creating probe informers")
-	}
-
 	if c.scrapeConfigSupported {
 		c.sconInfs, err = informers.NewInformersForResource(
```
Comment (xiu) [inline pkg/prometheus/agent/operator.go:142]: Do you want to centralize the creation of the ScrapeConfig informer as well or keep it as a follow up? (see #5545)

File: pkg/prometheus/server/operator.go [hunk 13]
```diff
@@ -210,91 +179,18 @@ func New(ctx context.Context, conf operator.Config, logger log.Logger, r prometh
...
-			resyncPeriod,
-			nil,
-		),
-		monitoringv1.SchemeGroupVersion.WithResource(monitoringv1.ProbeName),
-	)
-	if err != nil {
-		return nil, errors.Wrap(err, "error creating probe informers")
-	}
-
-	if c.scrapeConfigSupported {
```
Comment (xiu) [inline pkg/prometheus/server/operator.go:277]: Here we have a different behaviour. In the Agent operator we keep it but in the server one we remove it. Is that on purpose? see https://github.com/prometheus-operator/prometheus-operator/pull/5499/files#r1224188515

**Comment by mcbenjemaa** [comment]
i will rebase it.

**Comment by slashpai** [comment]
@mcbenjemaa  can you rebase?

**Comment by mcbenjemaa** [comment]
Yeah, I have to rebase.

File: pkg/prometheus/agent/operator.go (hunk 12 above)
Comment (mcbenjemaa) [reply pkg/prometheus/agent/operator.go:142]: followup make sense, as this PR become a huge, especially with rebasing 😄

File: pkg/prometheus/server/operator.go (hunk 13 above)
Comment (mcbenjemaa) [reply pkg/prometheus/server/operator.go:277]: don't know what do you mean.

Comment (xiu) [reply pkg/prometheus/server/operator.go:277]: Reviewing this comment, I actually added a comment on removed code, please disregard, thanks!

**Pull request closed without merging**